Sources/Plasma/NucleusLib/pnAsyncCoreExe/pnAceSocket.cpp

```diff
@@ -0,0 +1,658 @@
+        ) {
+            PerfAddCounter(kAsyncPerfSocketDisconnectBacklog, 1);
+
+            if (conn->fConnectionType) {
+                LogMsg(
+                    kLogPerf,
+                    "Backlog, c:{} q:{}, i:{}",
+                    conn->fConnectionType,
+                    currTimeMs - firstQueuedWrite->queueTimeMs,
+                    currTimeMs - conn->initTimeMs
+                );
+            }
+            AsyncSocketDisconnect((AsyncSocket) conn, true);
+            return false;
+        }
+    }
+
+    // If the last buffer still has space available then add data to it
+    if (!conn->fWriteOps.empty()) {
+        WriteOperation* op = conn->fWriteOps.back();
+        unsigned bytesLeft = std::min(op->fAllocSize - op->fNotify.bytes, bytes);
+        if (bytesLeft) {
+            PerfAddCounter(kAsyncPerfSocketBytesWaitQueued, bytesLeft);
+            memcpy(op->fNotify.buffer + op->fNotify.bytes, data, bytesLeft);
+            op->fNotify.bytes += bytesLeft;
+            data = (const uint8_t *)data + bytesLeft;
+            bytes -= bytesLeft;
+        }
+    }
+
+    if (bytes) {
+        // Allocate storage alongside the operation structure itself
+        unsigned bytesAlloc = std::max(bytes, kMinBacklogBytes);
+        auto membuf = new uint8_t[sizeof(WriteOperation) + bytesAlloc];
+        WriteOperation* op = new (membuf) WriteOperation;
+        conn->fWriteOps.emplace_back(op);
+
+        // TODO: asyncId
+        op->fAllocSize = bytesAlloc;
+        op->fNotify.asyncId = /* TODO */nullptr;
+        op->fNotify.buffer = membuf + sizeof(WriteOperation);
+        op->fNotify.bytes = bytes;
+        op->fNotify.bytesProcessed = 0;
+        op->queueTimeMs = TimeGetMs();
+        memcpy(op->fNotify.buffer, data, bytes);
+
+        PerfAddCounter(kAsyncPerfSocketBytesWaitQueued, bytes);
+    }
+
+    std::vector<asio::const_buffer> allWrites;
+    allWrites.reserve(conn->fWriteOps.size());
+    for (WriteOperation* op : conn->fWriteOps) {
+        if (op->fNotify.bytes - op->fNotify.bytesCommitted > 0) {
+            allWrites.emplace_back(op->AsBuffer());
+            op->fNotify.bytesCommitted = op->fNotify.bytes;
+        }
+    }
+    async_write(conn->fSock, allWrites, [conn, allWrites](const asio::error_code& err, size_t bytes) {
+        hsLockGuard(s_connectCrit);
+        while(bytes != 0) {
+            hsAssert(conn->fWriteOps.size() > 0, "buffer mismatch");
+            WriteOperation *op = conn->fWriteOps.front();
+            
+            size_t opBytesWritten = std::min(bytes, (size_t) op->fNotify.bytes - op->fNotify.bytesProcessed);
+            op->fNotify.bytesProcessed += opBytesWritten;
+            bytes -= opBytesWritten;
+            if  (op->fNotify.bytes == op->fNotify.bytesProcessed) {
+                conn->fWriteOps.pop_front();
+                
+                // callback notification procedure if requested
+                if (0) {

```

Comment:
This is the weird part. The async path checks if it needs to call a notification callback, and then in response to that callback, can soft disconnect.

In Windows, there is a op->notify boolean that controls if the callback should be called. The weird thing is - I can't find any place that is set to true. SocketQueueAsyncWrite even directly sets it to false.

I could be very wrong here - and something I haven't seen could be going down the soft disconnect path. The ASIO version doesn't have the notify boolean right now, so I'm just disabling the entire path.

Reply:
I'm absolutely not an expert on this low-level IO code, but it looks like most uses of `Nt::Operation.notify` were in `pnAceNtFile`, which H'uru removed way back in 2012 (d1da655fb2524fa6d45f8afc6df39da5e71cd4ee). One more dead use in `pnAceNtSocket` was removed in 2021 (e8b651d13a54c97a13f59f01a95f17231fd0e8f9).

In the current code, `op->notify` it's only set to `true` once at the start in `AsyncSocketConnect`, and then (presumably always) changed to `false` in `SocketQueueAsyncWrite`, as you say. So there should be no way for the callback to get called here. Even if it does get called, all code that handles the `kNotifySocketWrite` event looks like this:

```cpp
        case kNotifySocketWrite:
            // No action
        break;
```

So in my non-expert opinion, this code looks quite dead.

Reply:
I can remove this section that looks for a callback, and then add an assert on the disconnect function if anything out there tries a soft disconnect. That would invalidate other large sections of code in this pull request that could be removed.

Not necessarily on topic - how many of these network changes have hit the OpenUru client? One risk of removing these sections is breaking integration paths with OpenUru. But looking at the previous structure, it also seems unlikely any of that would survive a transition to ASIO.

Reply:
None, and all indication is that none ever will.